File: benchmark/synduce/constraints/bst/check_included_comb.f
Inductive Tree = leaf {Int, Int} | node {Int, Int, Tree, Tree};

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if < a b then b else a;

tmin = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf {a, b} -> a
  | node {a, b, l, r} -> min a (min (f l) (f r))
  end
);

tmax = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf {a, b} -> a
  | node {a, b, l, r} -> max a (max (f l) (f r))
  end
);

is_bst = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    leaf {a, b} -> < a b
  | node {a, b, l, r} -> and (< a b) (and (and (> a (tmax l)) (< a (tmin r))) (and (f l) (f r)))
  end
);

@Input lo: Int;
@Input hi: Int;

inside = \a: Int. \b: Int. and (< lo a) (< b hi);

spec = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    leaf {a, b} -> inside a b
  | node {a, b, l, r} -> or (or (inside a b) (f l)) (f r)
  end
);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    leaf {a, b} -> leaf {a, b}
  | node {a, b, l, r} ->
    if and (> b hi) (< a lo) then
      node {a, b, l, f r} /*Avoid the recursion on the left branch*/
    else node {a, b, f l, f r}
  end
);

main = \t: Tree. if is_bst t then spec (target t) else false;

File: ml-benchmark/synduce/constraints/bst/check_included_comb.ml
@Input val lo: int
@Input val hi: int

type tree = Leaf of int * int | Node of int * int * tree * tree

let min a b = if a < b then a else b
let max a b = if a < b then b else a

let rec tmin t =
  match t with
  | Leaf (a, b) -> a
  | Node (a, b, l, r) -> min a (min (tmin l) (tmin r))

let rec tmax t =
  match t with
  | Leaf (a, b) -> a
  | Node (a, b, l, r) -> max a (max (tmax l) (tmax r))

let rec is_bst t =
  match t with
  | Leaf (a, b) -> a < b
  | Node (a, b, l, r) -> (a < b) && ((a > tmax l) && (a < tmin r) && (is_bst l) && (is_bst r))

let inside a b = (lo < a) && (b < hi)

let rec spec t =
  match t with
  | Leaf (a, b) -> inside a b
  | Node (a, b, l, r) -> (inside a b) || (spec l) || (spec r)

val target: tree -> tree compress
let rec target t =
  match t with
  | Leaf (a, b) -> Leaf (a, b)
  | Node (a, b, l, r) ->
    if (b > hi) && (a < lo) then
      Node (a, b, l, target r)
    else
      Node (a, b, target l, target r)

let program t = if is_bst t then spec (target t) else false